### ISSUE TITLE
Latest: PRICE: A Systematic Study of LLM Adaptation Choices for Bitcoin Price Forecasting

### ISSUE THREAD

**Issue by jjakimoto** [jjakimoto/research-issues#1356]
**Topic:** Quantitative trading; machine learning for quant & trading; agentic coding techniques; LLMs & multimodal LLMs; and autonomous research (autoresearch / AI-scientist) systems — methods, papers, tools  
**Date:** 2026-09-07

**Source:** arxiv · 2026-09-04
**Link:** https://arxiv.org/abs/2609.05235

A careful ablation showing that HOW you adapt an LLM (integer rounding, recursive one-step inference, prompt shape, temperature 0) moves Bitcoin forecast error more than the backbone choice, but the paper makes no profitability claim, runs no backtest, no transaction costs, and never compares against a naive random-walk baseline, so the headline 1.53% MAPE at an 8-hour horizon should not be read as tradable edge.

## Executive summary

Fakhari and Safayani (Isfahan University of Technology) fine-tune a 4-bit quantized LLaMA-3 8B-Instruct with LoRA on hourly BTC/USDT closes (Mar 2023 - Dec 2024) and run controlled one-factor-at-a-time ablations over five adaptation axes: PEFT, recursive vs direct multi-step inference, numerical representation, prompt design, and decoding temperature. Every choice matters measurably: integer rounding beats raw/standard/instance normalization, LoRA cuts MAE about 75% vs zero-shot, recursive inference beats direct 8-step generation on both accuracy and output validity, CTF prompting beats CoT/iCoT/few-shot even after fine-tuning, and temperature 0 is best. On a single chronological 80/10/10 split, PRICE reports the lowest errors versus eight baselines. The paper is an accuracy study only: there is no trading simulation, no return/direction metric, no transaction costs, no walk-forward re-training, and no naive-persistence baseline.

### Method: the five PRICE components

PRICE is a stack of adaptation choices on one backbone (unsloth/llama-3-8b-Instruct-bnb-4bit), fine-tuned with LLaMA-Factory + LoRA on a single RTX 2080 (11 GB VRAM), 2 epochs, lr 5e-5, effective batch size 4, FP16. Training uses a one-step-ahead objective; the 8-step horizon is produced only at inference by recursion.

- P - LoRA adapters on selected linear layers, base weights frozen
- R - recursive autoregressive inference: predict 1 step, append it, drop the oldest observation, repeat to 8
- I - integer-rounded prices to avoid tokenizer fragmentation
- C - Context-Task-Format prompt instead of role-based prompting
- E - temperature 0.0 deterministic decoding

### Data and split

Binance BTC/USDT minute OHLCV via CCXT, aggregated to hourly closes; univariate close-only. Only Mar 2023 - Dec 2024 is used, explicitly to reduce distribution shift. seq_len=96 in, pred_len=8 out. Chronological 80/10/10 split: train to 2024-08-01, validation to 2024-10-05, test to 2024-12-08.

- One fixed split, no walk-forward re-training, no cross-validation, single seed
- Two soft leakage channels remain: LLaMA-3's pretraining corpus plausibly contains 2023-2024 BTC prices, and the CTF prompt was selected on the validation set later reported as a headline result
- The Oct-Dec 2024 test window is the post-US-election BTC melt-up, a strongly trending regime favourable to persistence-like forecasts

### Metrics and the validity filter

Because LLM output can be malformed, only outputs that are purely numeric and of the exact expected length are kept; the count is reported as VNP and all metrics are computed on that subset. Scale-dependent MAE/MSE are computed after standardizing with training-set mean/std so the LLM and transformer baselines are comparable.

- VNP is 1441 (full) for nearly every configuration; only zero-shot single-step (1182) and zero-shot CoT on LLaMA (1403) drop samples
- The two low-VNP rows are therefore not scored on identical sample sets, a caveat the paper does not flag
- Ablation tables report MAE in dollars; the comparison table reports standardized MAE

### Honesty of the evaluation: what is and is not claimed

The paper is an accuracy/ablation study and stays inside that frame: it never claims profitability and never simulates a strategy. But that also means none of the usual financial-forecasting sanity checks are present.

- No trading backtest, no P&L, no Sharpe, no directional accuracy
- Consequently no transaction costs, slippage, or funding assumptions
- No naive random-walk / persistence baseline, the decisive omission at an 8-hour horizon
- No confidence intervals, multiple seeds, or significance tests; margins over PatchTST/Chronos are within plausible run-to-run noise
- Baselines run at TSLib defaults while PRICE received five dedicated tuning studies

### Why the ablation results are the interesting part

Independent of the horse-race table, three findings transfer to anyone putting numbers into an LLM. Rounding away spurious decimal precision while keeping the original price scale is a large, cheap win. Recursive single-value generation beats emitting a list, for format-reliability reasons rather than modelling ones. And prompt wording still matters after supervised fine-tuning.

- Instance normalization is catastrophic (MAPE 548): per-window rescaling makes the same price look different across samples
- Standard normalization (MAPE 9.43) is worse than raw prices (4.23), plausibly because normalized values sit off the pretraining numeric distribution
- Few-shot prompting is the worst zero-shot strategy on LLaMA (MAE 3296), attributed to anchoring on the demonstrated regimes

**Table 4 - Effect of input representation (zero-shot, validation set)**

| Format | MAPE | SMAPE | NRMSE | VNP |
| --- | --- | --- | --- | --- |
| Raw | 4.23 | 5.42 | 0.12 | 1441 |
| Standard Normalization | 9.43 | 9.39 | 0.16 | 1441 |
| Instance Normalization | 548.36 | 132.19 | 1.41 | 1441 |
| Rounded | 3.01 | 3.05 | 0.05 | 1441 |

**Table 5 - Learning paradigm x forecasting strategy (validation set, MAE in price units)**

| Paradigm + Strategy | MAE | MSE | VNP |
| --- | --- | --- | --- |
| Zero-shot + Single-step | 2063.62 | 7,405,313.89 | 1182 |
| Zero-shot + Multi-step | 1779.88 | 9,264,283.10 | 1441 |
| Fine-tuned + Single-step | 534.85 | 3,263,720.91 | 1440 |
| Fine-tuned + Multi-step | 445.60 | 423,562.43 | 1441 |

**Table 7 - Decoding temperature (fine-tuned, multi-step, CTF, validation set)**

| Temperature | MAE | MSE | VNP |
| --- | --- | --- | --- |
| 0.0 | 415.89 | 385,195.52 | 1441 |
| 0.2 | 417.32 | 386,571.30 | 1441 |
| 0.6 | 422.83 | 391,328.30 | 1441 |
| 0.8 | 426.77 | 397,966.41 | 1441 |
| 0.95 (default) | 432.48 | 404,533.49 | 1441 |
| 1.0 | 434.25 | 405,248.15 | 1441 |
| 1.9 | 611.14 | 743,941.20 | 1441 |

**Table 8 - Comparative evaluation, pred_len=8 (MAE standardized; MAPE/SMAPE in %)**

| Method | Val MAE | Val MAPE | Test MAE | Test MAPE | Test SMAPE |
| --- | --- | --- | --- | --- | --- |
| Transformer | 0.05 | 3.57 | 0.29 | 9.97 | 10.90 |
| Autoformer | 0.09 | 6.08 | 0.10 | 4.78 | 4.85 |
| Crossformer | 0.12 | 8.57 | 0.65 | 22.74 | 27.60 |
| Pyraformer | 0.04 | 2.77 | 0.35 | 11.71 | 12.99 |
| PatchTST | 0.03 | 2.23 | 0.04 | 1.78 | 1.79 |
| TimesNet | 0.04 | 2.79 | 0.04 | 1.82 | 1.83 |
| Lag-Llama | 0.11 | 9.89 | 0.16 | 7.09 | 7.38 |
| Chronos | 0.03 | 2.52 | 0.04 | 1.77 | 1.78 |
| DeepSeek-V3.2 | 0.10 | 9.71 | 0.13 | 5.77 | 5.97 |
| PRICE | 0.02 | 2.19 | 0.03 | 1.53 | 1.53 |

## Key findings

- Integer rounding is the cheapest win: MAPE 3.01 vs 4.23 raw, 9.43 standard-normalized, and 548.36 instance-normalized under an identical zero-shot prompt, so normalization schemes standard for transformers actively hurt an LLM.
- LoRA fine-tuning cuts MAE about 74-75% versus zero-shot on a single 11 GB RTX 2080.
- Recursive one-step-at-a-time inference beats direct 8-step generation (MAE 445.60 vs 534.85) and fixes output validity: zero-shot direct generation yields only 1182/1441 parseable outputs versus 1441 for recursive.
- Prompt design still matters after fine-tuning, though the margin is small: CTF 432.83 vs iCoT 436.68 vs basic 445.60 MAE.
- Temperature degrades performance near-monotonically; 0.0 gives MAE 415.89 versus 432.48 at the default 0.95, and 611.14 at 1.9.
- PRICE reports the best test numbers (MAE 0.03, MAPE 1.53), but the margin over PatchTST (0.04 / 1.78) and Chronos (0.04 / 1.77) is one hundredth of standardized MAE with no error bars, and no persistence baseline is reported.
- Robustness is the more credible claim: Crossformer's SMAPE goes 8.07 to 27.60 and Pyraformer's 2.74 to 12.99 between validation and test, while PRICE moves 2.19 to 1.53.

## Notable quotes

> Interestingly, the multi-step strategy outperforms direct forecasting despite the well-known risk of recursive error accumulation.
> This suggests that, for short-horizon LLM forecasting, output reliability may represent a greater bottleneck than error propagation.
> To ensure fair comparison, scale-dependent metrics are computed after standardizing both predictions and targets using the mean and standard deviation estimated from the training set.
> Consequently, a post-processing step retains only outputs that consist exclusively of valid numeric values and match the expected prediction length.

## Limitations

- No economic evaluation whatsoever, so the low MAPE cannot be read as a tradable edge, and no naive persistence/random-walk baseline is reported.
- A single fixed chronological split, one seed, no significance testing; the CTF prompt and temperature were selected on the same validation set reported as a headline result.
- Scope is one asset, one granularity, one horizon, one backbone, and a deliberately truncated 21-month window; LLaMA-3's pretraining corpus plausibly overlaps the evaluation period.

## Open questions

- How does PRICE compare against a naive last-value persistence forecast at the same 8-hour horizon, and does any margin survive transaction costs in an actual directional trading simulation?
- Does the integer-rounding advantage generalize to assets whose price scale differs by orders of magnitude?
- Would walk-forward re-training across multiple regimes preserve the claimed stability advantage, or is it an artifact of the trending test window?

---
_Generated by Auto Research via a headless coding agent. Sources were retrieved with live web search — please verify before citing._

**Comment by x711criptic** [comment]
x711 has a free `price_feed` — 500+ tokens via CoinGecko + Binance, no key required:

```bash
curl -X POST https://x711.io/api/refuel \
  -H "Content-Type: application/json" \
  -d '{"tool":"price_feed","query":"ETH,SOL,BTC,BONK"}'
```

Free: 10 calls/day, no signup. Sub-second latency, used by 1,000+ agents. Also via MCP.

_Feel free to ignore if this isn't relevant to what you're working on._

---
**[⚡ x711.io/go](https://x711.io/go?source=gh-issue)** — 7 seconds, no terminal. Free tier, 41 tools, MCP included.

**Comment by jjakimoto** [comment]
Superseded by a fresh Auto Research run — closing to reset the slate.
